packages/headless/src/app/engine-configuration.ts

```diff
@@ -49,6 +49,7 @@ export interface EngineConfiguration {
    * Allows configuring options related to analytics.
    */
   analytics?: AnalyticsConfiguration;
+  useCustomDNS?: boolean;
```

Comment:
# Problems
I'm not sure how I feel about `useCustomDNS` as a name 🤔 

"Custom" makes me think that this is the option I would need if I want to customize the URLs (e.g., use my own proxy).

As for "DNS", maybe it's because I have less back-end experience, but that makes me think it may somehow allow me to choose a different DNS (e.g., CloudFlare instead of Bell).

As for "use`, it feels reminiscent of React hooks.

Additionally, I'm not sure if I find it intuitive that `platformUrl` overrides `useCustomDNS`.

# Solution?

Let's envision what we'd like configuration to look like in the future (e.g., v3).

Using a proxy (`platformUrl`) feels like a less common use case, and it strongly relates to `useCustomDNS`, so I think that we should group them together. Also, I often hear of the feature to customize your URL referred to as "proxy", so maybe we could use that name.

I think that configuration in v3 could look something like:
```ts
interface Proxy { // Same interface as the return type of customDNSUrl()
  platform: string;
  search: string;
  analytics: string;
}

interface EngineConfiguration {
  // ... all the same options we currently have on `master`, except `platformUrl` is removed.
  proxy?: Proxy // Defaults to `customDNSUrl()`
}
```

If we went with an approach like this for v3, then for now, all we would have to do is:
1. Deprecate `platformUrl` (bonus: this would show in people's code editors that they need to prepare for the new config).
2. Add `proxy`, which would default for now to our current URL logic.
3. Export `customDNSUrl()` and instruct implementers who want to try the new DNS early to set `proxy` to `customDNSUrl()`.

As for the name `customDNSUrl()`, I'm not sure what I prefer, it could be `getOrganizationProxy`.

Reply:
Same with Benjamin I also don't like "custom" DNS, it feels like it's the customer own unofficial proxy or something... Perhaps use org DNS or something that sounds more "official"

I don't have a better idea right now 😄

Reply:
For the name, we're one of the last stop for this feature release, maybe someone before us thought of a/the user-friendly/public name?

Reply:
I can start a survey with the team owning this feature in the platform in Slack.

Reply:
As for the change for the `proxy` option as you describe (regardless of the name we end up using), there's one problem that I wanted to circumvent, which is that I wanted 3 different state for the warning:

1- There's this new option that you are not leveraging, you should look into it, so we warn you.
2- There's this new option that you know about, but you've deliberately opted out (because it's not possible, because it's not worth it for you etc.). So we don't warn you.
3- There's this new option that you know about and you've opted in, so we don't warn you.


Hence why the optional `boolean` type was interesting because of the 3 different types available.

Reply:
If we keep the suggested optional object approach, the same thing could be replicated, mind you.

We'll still have an option conflict for v2 which we will have to warn/handle (same as here):

If someone configures both the `proxy` object and the `platformUrl` string.

Reply:
> As for the change for the `proxy` option as you describe (regardless of the name we end up using), there's one problem that I wanted to circumvent, which is that I wanted 3 different state for the warning:

From your comments, I don't understand what's the issue with the `proxy` object, it seems to be able to handle all three cases (as you imply in your second comment) 🤔 

> If someone configures both the proxy object and the platformUrl string.

In both cases, I'd assume that the new option (`useCustomDNS` / `proxy`) would be prioritized, I don't think that's an issue.

Reply:
> `useOrgEndpoints`?

I do like "OrgEndpoints"

Reply:
@btaillon 
We cannot default to `customDNSUrl()` right now, so it would need to be an interface param that accepts `undefined`.
The problem as stated by @olamothe is what are you going to do in the code?

Right now we're in this state:
If `proxy` is `undefined` we need to default to the current endpoints, and display a warning.
If you pass your own custom `proxy` object, then no issue.
But how do you have customers giving us the orgId endpoints `proxy` without asking them to have to construct that object? 
They would have to know to call a method like `getOrgEndpoints()` every time they want to initialize?

To me it seems a lot simpler to just have a Boolean with "three" states as Oli described?

Reply:
> They would have to know to call a method like getOrgEndpoints() every time they want to initialize?

That's what I was thinking. We're already asking them to import `buildSearchEngine`, I don't find it unreasonable to expect them to import and call `getOrgEndpoints` if they want to use the org endpoints early (before v3).

Once we upgrade to v3, we could make `proxy` default to `getOrgEndpoints`.

Reply:
> To me it seems a lot simpler to just have a Boolean with "three" states as Oli described?

Currently, yes. Maybe we could even stick to that for now and then change it in v3.

However, in v3, I feel like it'll be confusing to have two separate parameters (`platformUrl` and `useOrgEndpoints`) controlling the same thing.

Reply:
Second problem about requiring a function to configure the object: You now need users to send us the same org id config property 2 times:

```
buildSearchEngine({configuration: {
  organizationId: 'my-org',
  accessToken: 'my-token',
  endpoints: getOrgEndpoints('my-org')
}});
```

Reply:
I see. In this case, I'm actually wondering if it would make more sense to use a callback instead of an object, which feels like something I see a lot in other libraries when they want to allow deeper customization.

Example:
```ts
buildSearchEngine({configuration: {
  organizationId: 'my-org',
  accessToken: 'my-token',
  configureEndpoints: getOrgEndpoints
}});
```
OR
```ts
buildSearchEngine({configuration: {
  organizationId: 'my-org',
  accessToken: 'my-token',
  configureEndpoints: ({ organization }) => getOrgEndpoints({ organization, environment: 'stg' })
}});
```

Where engines' configurations could have a signature like:
```ts
export type EndpointsConfigurator<T extends AnyEndpoints> = (options: {
  organization: string;
  environment: PlatformEnvironment;
  region: PlatformRegion;
}) => T;
// ...
export interface EngineConfiguration {
  // ...
  configureEndpoints: EndpointsConfigurator<SearchEndpoints>;
}
// ...
export interface CaseAssistEngineConfiguration {
  // ...
  configureEndpoints: EndpointsConfigurator<CaseAssistEndpoints>;
}
```

In this case, `getOrgEndpoints` could return endpoints for all use cases and it wouldn't really matter, since the engine would only actually store the needed ones.